### PR TITLE
Adding support for ASan (Windows/MSVC) and fixing AzCore

### PR DESCRIPTION
**Pull request by amznestebanpapp**
This PR adds support to build with ASan (Address Sanitizer) enabled (Windows/MSVC)

ASan is a tool to detect many memory issues, details: https://docs.microsoft.com/en-us/cpp/sanitizers/asan?view=msvc-160

This PR also fixes all issues exposed by AzCore.Tests

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/IO/Path/Path.inl [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 #include <AzCore/std/containers/array.h>
 #include <AzCore/std/string/wildcard.h>
+#include <AzCore/Casting/numeric_cast.h>
```
Comment (amznestebanpapp) [inline Code/Framework/AzCore/AzCore/IO/Path/Path.inl:13]: its being used on line 225

File: Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp [hunk 2]
```diff
@@ -17,15 +17,23 @@
...
+#if AZCORE_SYSTEM_ALLOCATOR == AZCORE_SYSTEM_ALLOCATOR_HPHA
+    #include <AzCore/Memory/HphaSchema.h>
+#elif AZCORE_SYSTEM_ALLOCATOR == AZCORE_SYSTEM_ALLOCATOR_MALLOC
+    #include <AzCore/Memory/MallocSchema.h>
+#elif AZCORE_SYSTEM_ALLOCATOR == AZCORE_SYSTEM_ALLOCATOR_HEAP
+    #include <AzCore/Memory/HeapSchema.h>
 #else
-#   include <AzCore/Memory/HeapSchema.h>
+    #error "Invalid allocator selected for SystemAllocator"
 #endif
```
Comment (amznestebanpapp) [inline Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp:37]: the HphaSchema doesnt play well with ASan because most of the memory is not returned. The hooks to ASan happen at the OS allocation level. Changed this to be able to pick the schema to use for the SystemAllocator (when ASan is enabled it will use `MallocSchema`)

File: Code/Framework/AzCore/Tests/Math/SfmtTests.cpp [hunk 3]
```diff
@@ -27,8 +27,8 @@ namespace UnitTest
         void SetUp() override
         {
             AllocatorsFixture::SetUp();
-            array1 = (AZ::u64*)azmalloc(sizeof(AZ::u64) * 2 * (BLOCK_SIZE / 4), AZStd::alignment_of<AZ::u64>::value);
-            array2 = (AZ::u64*)azmalloc(sizeof(AZ::u64) * 2 * (10000 / 4), AZStd::alignment_of<AZ::u64>::value);
+            array1 = (AZ::u64*)azmalloc(sizeof(AZ::u64) * 2 * (BLOCK_SIZE / 4), AZStd::alignment_of<AZ::Vector4>::value);
+            array2 = (AZ::u64*)azmalloc(sizeof(AZ::u64) * 2 * (10000 / 4), AZStd::alignment_of<AZ::Vector4>::value);
```
Comment (amznestebanpapp) [inline Code/Framework/AzCore/Tests/Math/SfmtTests.cpp:31]: this was producing a lot of alignment errors when operating over that memory with SIMD

File: Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsTypes.h [hunk 4]
```diff
@@ -13,6 +13,7 @@
 #include <AzCore/std/containers/variant.h>
 #include <AzCore/std/containers/vector.h>
 #include <AzCore/std/smart_ptr/shared_ptr.h>
+#include <AzCore/Casting/numeric_cast.h>
```
Comment (amznestebanpapp) [inline Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsTypes.h:16]: used on line 110

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp [hunk 5]
```diff
@@ -587,7 +604,9 @@ namespace UnitTest
         AZ::NameDictionary::Create();
 
         // 3 threads per name effectively makes two readers and one writer (the first to run will write in the dictionary)
-        RunConcurrencyTest<ThreadCreatesOneName>(AZ_TRAIT_UNIT_TEST_NAME_COUNT, 3);
+        // Using AZ_TRAIT_UNIT_TEST_NAME_COUNT / 2 since the following line generates 3000 threads and that triggers ASan failures,
+        // likely because of https://devblogs.microsoft.com/oldnewthing/20050729-14/?p=34773
+        RunConcurrencyTest<ThreadCreatesOneName>(AZ_TRAIT_UNIT_TEST_NAME_COUNT / 2, 3);
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: I think we should cut this trait down even more.
I would make the thread count 128 threads max.

File: Code/Legacy/CryCommon/LegacyAllocator.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
```
Comment (lumberyard-employee-dm) [inline Code/Legacy/CryCommon/LegacyAllocator.cpp:9]: `#pragma once` has to be removed from the file(Clang will fail to build)

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Code/Legacy/CryCommon/LegacyAllocator.cpp (hunk 6 above)
Comment (amznestebanpapp) [reply Code/Legacy/CryCommon/LegacyAllocator.cpp:9]: good catch, copy paste fail

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp (hunk 5 above)
Comment (amznestebanpapp) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: I dont disagree... just concerned what the intention was here... 
I will leave it up to the test author

File: cmake/Configurations.cmake [hunk 7]
```diff
@@ -20,73 +20,80 @@ include_guard(GLOBAL)
...
+        set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} ${LINK_STR}" PARENT_SCOPE)
+    endif()
 
-        set(CMAKE_EXE_LINKER_FLAGS ${LINK_EXECUTABLE_OPTIONS} PARENT_SCOPE)
+    if(ly_append_configurations_options_LINK_SHARED)
+        string(REPLACE ";" " " LINK_STR "${ly_append_configurations_options_LINK_SHARED}")
+        set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${LINK_STR}" PARENT_SCOPE)
     endif()
 
     foreach(conf IN LISTS CMAKE_CONFIGURATION_TYPES)
```
Comment (lumberyard-employee-dm) [inline cmake/Configurations.cmake:99]: Technically we could consolidate the pre-loop code with the for loop if we prepend an empty item to a list and iterator over that
```suggestion
    set(configuration_items ";${CMAKE_CONFIGURATION_TYPES}")
    foreach(conf IN LISTS configuration_items)
        unset(UCONF)
        if(conf)
           string(TOUPPER "_${conf}" UCONF)
        endif()
```

File: Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp [hunk 8]
```diff
@@ -185,8 +185,25 @@ namespace AZ
             return;
         }
 
+        // Get the hash before locking since another thread could be deleting the object within the lock
+        Internal::NameData::Hash hash = nameData->GetHash();
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp:189]: Something still doesn't seem right here.

`nameData` can still be deleted before even getting to this line

Therefore I think the right solution is to retrieve the hash value, before decrementing the NameData count in the [NameData::release()](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Name/Internal/NameData.cpp#L40)

Here is the scenario
- T1, gets into TryReleaseName, but has not looked up the hash yet
- T2 gets into MakeName, acquires the lock, returns a new Name that increments the counter
- T2 Name decrements the counter, gets into TryReleaseName
- T2 looks up the hash value
- T2 gets the acquires the lock
- T2 Deletes the NameData
- T1 **now looks up the hash** and accesses invalid memory due to nameData being deleted

**Review by lumberyard-employee-dm — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp (hunk 8 above)
Comment (amznestebanpapp) [reply Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp:189]: you are right, let me edit it

File: cmake/Configurations.cmake (hunk 7 above)
Comment (amznestebanpapp) [reply cmake/Configurations.cmake:99]: good idea, will do it in a different PR

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: @santorac What are you thoughts on this, since your are the Test author.

1500 threads is still a lot and assuming a 1 Megabyte Stack per thread, that uses about 1.5 GiB of ram.
Considering that we consider the minimum [requirements](https://o3de.org/docs/welcome-guide/requirements/) for the engine 8GiB of CPU ram, that is still about 18% of the ram for the stack in these test..

Comment (amznestebanpapp) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: @santorac if you are ok with @lumberyard-employee-dm suggestion, I am happy to change it.
Note that the `3` there is a multiplier in the number of threads, so I would change `AZ_TRAIT_UNIT_TEST_NAME_COUNT` to something like `40`

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp (hunk 2 above)
Comment (jeremyong) [reply Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp:37]: Did you get a chance to look at the asan interface? I think you can unpoison regions or mark regions as being freed by some other mechanism.

https://github.com/llvm-mirror/compiler-rt/blob/master/include/sanitizer/asan_interface.h

**Comment by jeremyong-az** [comment]
Thanks a ton for doing this. I'll be sure to be running with it on frequently in fact, when the perf is not needed.

We should probably file a downstream issue to ensure that our data structures (vectors/maps/etc.) poison memory regions properly so that reserved/slack memory is considered "free" with respect to the ASAN runtime.

File: Code/Legacy/CryCommon/LegacyAllocator.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+#include <CryCommon/LegacyAllocator.h>
+
+namespace AZ
+{
+    LegacyAllocator::pointer_type LegacyAllocator::Allocate(size_type byteSize, size_type alignment, int flags, const char* name, const char* fileName, int lineNum, unsigned int suppressStackRecord)
+    {
+        if (alignment == 0)
+        {
+            // Some STL containers, like std::vector, are assuming a specific minimum alignment. seems to have a requirement
```
Comment (mbalfour-amzn) [inline Code/Legacy/CryCommon/LegacyAllocator.cpp:17]: Not quite sure what this comment is saying with the "seems to have a requirement" at the end.  Maybe rephrase to "Some STL containers, like std::vector, seem to have a requirement where a specific minimum alignment will be chosen when the alignment is set to 0"?

File: Code/Legacy/CryCommon/LegacyAllocator.cpp [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        AZ_MEMORY_PROFILE(ProfileDeallocation(ptr, byteSize, alignment, nullptr));
+        m_schema->DeAllocate(ptr, byteSize, alignment);
+    }
+
+    // Realloc with file/line, because Cry uses realloc(nullptr) and realloc(ptr, 0) to mimic malloc/free
+    LegacyAllocator::pointer_type LegacyAllocator::ReAllocate(pointer_type ptr, size_type newSize, size_type newAlignment, [[maybe_unused]] const char* file, [[maybe_unused]] const int line)
+    {
+        if (newAlignment == 0)
+        {
+            // Some STL containers, like std::vector, are assuming a specific minimum alignment. seems to have a requirement
```
Comment (mbalfour-amzn) [inline Code/Legacy/CryCommon/LegacyAllocator.cpp:42]: Same as above - not sure what this comment means.

File: Code/Legacy/CryCommon/LegacyAllocator.cpp [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        AZ_MEMORY_PROFILE(ProfileDeallocation(ptr, 0, 0, nullptr));
+        Base::DeAllocate(ptr, byteSize, alignment);
+    }
+
+    LegacyAllocator::pointer_type LegacyAllocator::ReAllocate(pointer_type ptr, size_type newSize, size_type newAlignment)
+    {
+        if (newAlignment == 0)
+        {
+            // Some STL containers, like std::vector, are assuming a specific minimum alignment. seems to have a requirement
```
Comment (mbalfour-amzn) [inline Code/Legacy/CryCommon/LegacyAllocator.cpp:66]: Same here...

**Review by mbalfour-amzn — APPROVED** [review]

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp (hunk 5 above)
Comment (spham-amzn) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: Keep in mind the 'AZ_TRAIT_UNIT_TEST_NAME_COUNT' is a PAL trait and was intended to be adjusted anyways per platform. I would think a better value for this should reflect how many threads the engine might use in normal circumstances.

**Comment by amznestebanpapp** [comment]
> memory

yeah, definitely need to add that at some point. Currently the MSVC implementation of ASan doesn't have an API, so I can't add it to containers The clang version does have an API. The MSVC implementation integrates with VS so it was more straightforward/practical.

File: Code/Legacy/CryCommon/LegacyAllocator.cpp (hunk 9 above)
Comment (amznestebanpapp) [reply Code/Legacy/CryCommon/LegacyAllocator.cpp:17]: I actually just moved this code from the header to a cpp, will rephrase to your suggestion

File: Code/Legacy/CryCommon/LegacyAllocator.cpp (hunk 10 above)
Comment (amznestebanpapp) [reply Code/Legacy/CryCommon/LegacyAllocator.cpp:42]: same, will follow your suggestion

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp (hunk 5 above)
Comment (amznestebanpapp) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: true, if @santorac is ok, I wont change `AZ_TRAIT_UNIT_TEST_NAME_COUNT` but just change the value that gets passed in this case.
Problem is that in this case threads are used to cause collisions on the use of Name variables and dictionary access. So the number is not really tied to how many threads should be running in the engine, its trying to force more race/contention conditions to make threading problems more likely to occur.

File: Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp [hunk 12]
```diff
@@ -179,13 +179,22 @@ namespace AZ
...
+            // T1 gets the lock, goes to the compare_exchange if and has a counter of 0, deletes
+            // Then T2 continues, gets the lock and crashes because nameData was deleted
             return;
         }
 
-        AZStd::unique_lock<AZStd::shared_mutex> lock(m_sharedMutex);
+        Internal::NameData* nameData = dictIt->second; // restore the pointer in case the intrusive ptr was already assigned by other thread
 
         // Check m_hashCollision again inside the m_sharedMutex because a new collision could have happened
         // on another thread before taking the lock.
```
Comment (santorac) [inline Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp:200]: This comment is stale ("again" is no longer applicable), needs to be revised.

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp (hunk 5 above)
Comment (santorac) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: Though my name is on this code, I'm not the original author (he left Amazon some time ago). I was responsible for finishing it up and making some final improvements. I concur that AZ_TRAIT_UNIT_TEST_NAME_COUNT is used for more than just driving thread counts, so I don't think we should change it for the sake of this situation; rather we should adjust the use of RunConcurrencyTest. 

I recommend creating a second trait AZ_TRAIT_UNIT_TEST_MAX_THREAD_COUNT = 2000 (or less if you want). Then modify the name loop in RunConcurrencyTest to go up to min(AZ_TRAIT_UNIT_TEST_MAX_THREAD_COUNT / threadsPerName, nameCount).

**Review by santorac — COMMENTED** [review]
I only reviewed the Name related code.

File: Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp [hunk 13]
```diff
@@ -179,13 +179,22 @@ namespace AZ
...
+            // T1, gets into TryReleaseName
+            // T2 gets into MakeName, acquires the lock, returns a new Name that increments the counter
+            // T2 deletes the Name decrements the counter, gets into TryReleaseName
+            // T1 gets the lock, goes to the compare_exchange if and has a counter of 0, deletes
+            // Then T2 continues, gets the lock and crashes because nameData was deleted
             return;
         }
 
-        AZStd::unique_lock<AZStd::shared_mutex> lock(m_sharedMutex);
+        Internal::NameData* nameData = dictIt->second; // restore the pointer in case the intrusive ptr was already assigned by other thread
```
[the remark this answers is not in the document]
Comment (amznestebanpapp) [reply Code/Framework/AzCore/AzCore/Name/NameDictionary.cpp:197]: true, that was a leftover of an intermediate change, will remove

File: Code/Framework/AzCore/Tests/Name/NameTests.cpp (hunk 5 above)
Comment (amznestebanpapp) [reply Code/Framework/AzCore/Tests/Name/NameTests.cpp:609]: will remove `AZ_TRAIT_UNIT_TEST_NAME_COUNT` because is just used in these tests and one other test where I can just set the value there
The trait is also in the wrong place, it was added to AzTest when the tests are in AzCore.Tests

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp (hunk 2 above)
Comment (amznestebanpapp) [reply Code/Framework/AzCore/AzCore/Memory/SystemAllocator.cpp:37]: that is clang, MSVC doesnt have that interface yet and this implementation is just for MSVC

**Comment by jeremyong** [comment]
Note for @amznestebanpapp , I believe ASAN on MSVC does in fact support the region poisoning/unpoisoning interface already. See https://docs.microsoft.com/en-us/cpp/sanitizers/asan-runtime?view=msvc-160#poisoning

**Comment by amznestebanpapp** [comment]
> Note for @amznestebanpapp , I believe ASAN on MSVC does in fact support the region poisoning/unpoisoning interface already. See https://docs.microsoft.com/en-us/cpp/sanitizers/asan-runtime?view=msvc-160#poisoning

awesome, will add it to my list!
